oauthenticator/oauth2.py

```diff
@@ -277,8 +288,45 @@ def _token_url_default(self):
     def _userdata_url_default(self):
         return os.environ.get("OAUTH2_USERDATA_URL", "")
 
+    username_claim = Unicode(
+        config=True,
+        help="""Field in userdata reply to use for username
+        The field in the userdata response from which to get the JupyterHub username.
+        Examples include: email, username, nickname
+
+        What keys are available will depend on the scopes requested and the authenticator used.
+        """,
+    )
+
+    @default("username_claim")
+    def _username_claim_default(self):
+        return "username"
+
+    # Enable refresh_pre_spawn by default if self.enable_auth_state
+    @default("refresh_pre_spawn")
+    def _refresh_pre_spawn(self):
+        if self.enable_auth_state:
+            return True
+
     logout_redirect_url = Unicode(config=True, help="""URL for logging out of Auth0""")
 
+    # Originally a GenericOAuthenticator only trait
+    userdata_params = Dict(
+        help="Userdata params to get user data login information"
+    ).tag(config=True)
+
+    # Originally a GenericOAuthenticator only trait
+    userdata_token_method = Unicode(
+        os.environ.get("OAUTH2_USERDATA_REQUEST_TYPE", "header"),
+        config=True,
+        help="Method for sending access token in userdata request. Supported methods: header, url. Default: header",
+    )
+
+    # Originally a GenericOAuthenticator only trait
+    extra_params = Dict(
```

Comment:
Should we take this opportunity to call it `token_params` to be consistent with `userdata_params`?

```suggestion
    token_params = Dict(
```